### PR TITLE
Use `Write.TxOut` in `ErrBalanceTxInputResolutionConflicts`.

### PR DESCRIPTION
**Pull request by jonathanknowles**
## Issue

ADP-3184

## Description

This PR adjusts `ErrBalanceTxInputResolutionConflicts` to use `Write.TxOut` instead of the wallet primitive `TxOut` type.

### CODE REVIEW

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 1]
```diff
@@ -991,8 +991,18 @@ decodeSealedTxSpec = describe "SealedTx serialisation/deserialisation" $ do
...
-feeEstimationRegressionSpec = describe "Regression tests" $ do
+feeEstimationRegressionSpec :: AnyRecentEra -> Spec
+feeEstimationRegressionSpec (AnyRecentEra era) =
+    case era of
+        RecentEraConway ->
+            feeEstimationRegressionSpecInner @Cardano.ConwayEra era
+        RecentEraBabbage ->
+            feeEstimationRegressionSpecInner @Cardano.BabbageEra era
+
+feeEstimationRegressionSpecInner
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: It seems unnecessary to define `feeEstimationRegressionSpecInner` separately from `feeEstimationRegressionSpec` to me. I'll create another PR

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs [hunk 2]
```diff
@@ -902,14 +904,16 @@ balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment
         -> ExceptT (ErrBalanceTx era) m ()
     guardWalletUTxOConsistencyWith u' = do
         let W.UTxO u = toWalletUTxO (recentEra @era) $ fromCardanoApiUTxO u'
-        let conflicts = lefts $ flip map (Map.toList u) $ \(i, o) ->
+        let conflicts = lefts $ flip map (Map.toList u) $ \(i, out1) ->
```
Comment (Anviking) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: Changing `o` to `out` breaks the visual similarity with `i`. Is this really worth it?

File: lib/wallet/test/unit/Cardano/WalletSpec.hs [hunk 3]
```diff
@@ -759,11 +760,16 @@ walletListsOnlyRelatedAssets txId txMeta =
 -- 1. There is no coin selection with a fee above the estimated maximum.
 -- 2. The minimum estimated fee is no greater than the maximum estimated fee.
 -- 3. Around 10% of fees are below the estimated minimum.
-prop_calculateFeePercentiles :: NonEmptyList (Maybe Coin) -> Property
-prop_calculateFeePercentiles (NonEmpty coins) =
+prop_calculateFeePercentiles
+    :: Write.AnyRecentEra
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/WalletSpec.hs:764]: 👍

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 4]
```diff
@@ -1002,7 +1013,7 @@ feeEstimationRegressionSpec = describe "Regression tests" $ do
                     { requiredCost = Ledger.Coin $ intCast requiredCostLovelace
                     , shortfall = Ledger.Coin 100_000
                     }
-        result <- runExceptT (calculateFeePercentiles estimateFee)
+        result <- runExceptT (calculateFeePercentiles @_ @era estimateFee)
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: Not specific to this PR but just thinking about our direction: Maybe the need for additional type-application is a bad sign 🤔

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs (hunk 1 above)
Comment (Anviking) [reply lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: https://github.com/cardano-foundation/cardano-wallet/pull/4207

**Review by Anviking — COMMENTED** [review]
lgtm modulo two comments

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs (hunk 2 above)
Comment (jonathanknowles) [reply lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: Here I was aiming for visual similarity between `out1` and `out2`:
```hs
                   Just out2 -> unless (out1 == out2) $ Left
                        ( toLedgerTxOut era out1
                        , toLedgerTxOut era out2
                        )
```

I can rename them to `o1` and `o2` if you prefer?

Comment (Anviking) [reply lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: I'd prefer sticking to `o`, `o'`, or possibly `o1` or `o2` if you want them to have the same length

Comment (Anviking) [reply lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: Although the added descriptiveness might be worth it as you have it 🤔 => 🤷‍♂️ approved

**Review by Anviking — APPROVED** [review]

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs (hunk 2 above)
Comment (jonathanknowles) [reply lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: Okay! Fixed in 6e3879e6acb184eaa40b20a057fbeae60266d760.

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: @Anviking Thanks for the suggestion! I've cherry-picked your suggested commit into this branch. 👍🏻

**Pull request merged**